Mathlib/GroupTheory/Perm/Support.lean

```diff
@@ -166,111 +146,188 @@ theorem Disjoint.mul_apply_eq_iff {σ τ : Perm α} (hστ : Disjoint σ τ) {a
+  obtain ⟨_, _, hxy, hf⟩ := h
+  exact ⟨_, _, hxy, hf.inv⟩
 
-end IsSwap
+theorem IsSwap.symm {f : Perm α} (h : f.IsSwap) : IsSwap f.symm := h.inv
 
-section support
+theorem IsSwap.inv_eq {f : Perm α} (h : f.IsSwap) : f⁻¹ = f := by
+  obtain ⟨_, _, _, hf⟩ := h
+  exact hf.inv_eq
 
-section Set
+theorem IsSwap.symm_eq {f : Perm α} (h : f.IsSwap) : f.symm = f := h.inv_eq
 
-variable (p q : Perm α)
+theorem IsSwap.apply_self {f : Perm α} {z : α}
+    (h : f.IsSwap) : f (f z) = z := by rw [← eq_inv_iff_eq, h.inv_eq]
 
-theorem set_support_inv_eq : { x | p⁻¹ x ≠ x } = { x | p x ≠ x } := by
-  ext x
-  simp only [Set.mem_setOf_eq, Ne]
-  rw [inv_def, symm_apply_eq, eq_comm]
+theorem IsSwap.mul_self {f : Perm α} (h : f.IsSwap) : f * f = 1 := by
+  rw [mul_eq_one_iff_inv_eq, h.inv_eq]
 
-theorem set_support_apply_mem {p : Perm α} {a : α} :
-    p a ∈ { x | p x ≠ x } ↔ a ∈ { x | p x ≠ x } := by simp
+theorem IsSwap.trans_self {f : Perm α} (h : f.IsSwap) : f.trans f = Equiv.refl α := h.mul_self
 
-theorem set_support_zpow_subset (n : ℤ) : { x | (p ^ n) x ≠ x } ⊆ { x | p x ≠ x } := by
-  intro x
-  simp only [Set.mem_setOf_eq, Ne]
-  intro hx H
-  simp [zpow_apply_eq_self_of_apply_eq_self H] at hx
+theorem IsSwap.ne_one {f : Perm α} (h : f.IsSwap) : f ≠ 1 := by
+  obtain ⟨_, _, hxy, hf⟩ := h
+  exact DFunLike.ne_iff.mpr ⟨_, Eq.trans_ne hf.2.2 hxy⟩
 
-theorem set_support_mul_subset : { x | (p * q) x ≠ x } ⊆ { x | p x ≠ x } ∪ { x | q x ≠ x } := by
-  intro x
-  simp only [Perm.coe_mul, Function.comp_apply, Ne, Set.mem_union, Set.mem_setOf_eq]
-  by_cases hq : q x = x <;> simp [hq]
+theorem IsSwap.of_subtype_isSwap {p : α → Prop} [DecidablePred p] {f : Perm (Subtype p)} :
+    f.IsSwap → (ofSubtype f).IsSwap := by
+  simp_rw [IsSwap, ← Subtype.coe_ne_coe]
+  exact fun ⟨_, _, hxy, H⟩ => ⟨_, _, hxy, H.of_subtype_isSwapOn⟩
 
-end Set
+theorem swap_isSwap [DecidableEq α] {x y : α} (h : x ≠ y) : IsSwap (swap x y) :=
+  ⟨x, y, h, swap_isSwapOn⟩
 
-variable [DecidableEq α] [Fintype α] {f g : Perm α}
+theorem isSwap_iff_exists_distinct_eq_swap [DecidableEq α] {f : Perm α} :
+    f.IsSwap ↔ ∃ x y, x ≠ y ∧ f = swap x y := by
+  simp_rw [IsSwap, isSwapOn_iff_eq_swap]
 
-/-- The `Finset` of nonfixed points of a permutation. -/
-def support (f : Perm α) : Finset α :=
-  univ.filter fun x => f x ≠ x
+theorem IsSwap.exists_distinct_eq_swap [DecidableEq α] {f : Perm α} (hf : f.IsSwap) :
+    ∃ x y, x ≠ y ∧ f = swap x y := isSwap_iff_exists_distinct_eq_swap.mp hf
 
-@[simp]
-theorem mem_support {x : α} : x ∈ f.support ↔ f x ≠ x := by
-  rw [support, mem_filter, and_iff_right (mem_univ x)]
+end IsSwap
 
-theorem not_mem_support {x : α} : x ∉ f.support ↔ f x = x := by simp
+section support
 
-theorem coe_support_eq_set_support (f : Perm α) : (f.support : Set α) = { x | f x ≠ x } := by
-  ext
-  simp
+variable {f g : Perm α}
+
+/-- The `Set` of nonfixed points of a permutation. -/
+def support (f : Perm α) : Set α := { x | f x ≠ x }
```

Comment:
As explained on Zulip, can you please keep the finset-valued support around (along with its API, which does **not** include the lemmas that mention it just to talk about its membership as there it can be replaced by the new `support`), maybe named `supportFinset`?

And to help with reviewing can you make a first PR that renames `support` to `supportFinset` everywhere? This will create a lot of conflicts with the current PR, but you can systematically solve them by taking the change from this PR (except for the ones that remove the `supportFinset` API)